inputremapper/injection/mapping_handlers/rel_to_rel_handler.py

```diff
@@ -0,0 +1,220 @@
+
+import asyncio
+from typing import Dict, Tuple, Optional
+
+import evdev
+from evdev.ecodes import (
+    EV_REL,
+    REL_WHEEL,
+    REL_HWHEEL,
+)
+
+from inputremapper import exceptions
+from inputremapper.configs.mapping import Mapping
+from inputremapper.event_combination import EventCombination
+from inputremapper.injection.global_uinputs import global_uinputs
+from inputremapper.injection.mapping_handlers.axis_transform import Transformation
+from inputremapper.injection.mapping_handlers.mapping_handler import (
+    MappingHandler,
+    HandlerEnums,
+    InputEventHandler,
+)
+from inputremapper.input_event import InputEvent
+from inputremapper.logger import logger
+
+# a trial-and-error value for the wheel speed
+# TODO move to preset config and use 16 as default
+WHEEL_FACTOR = 16
+
+MAX_REL_XY = 256
+
+
+def is_wheel(event) -> bool:
+    return event.type == EV_REL and event.code in (REL_WHEEL, REL_HWHEEL)
+
+
+# TODO
+"""
+00:16:22.695476 35088 Service ERROR preset.py:308: failed to Validate mapping for 2,0,1: 1 validation error for Mapping
+__root__
+  missing macro or key: the combination = <EventCombination (2, 0, 1)> specifies a key input, but no output macro or key is programmed (type=value_error)
+  
+  
+00:18:34.576837 35275 Service ERROR preset.py:308: failed to Validate mapping for 2,0,1: 2 validation errors for Mapping
+__root__
+  missing Argument: Mapping must either contain `output_symbol` or `output_type` and `output_code` (type=value_error)
+__root__
+  missing macro or key: the combination = <EventCombination (2, 0, 1)> specifies a key input, but no output macro or key is programmed (type=value_error)
+"""
+
+# TODO after "The device was not grabbed" message, reset the "Stop" button
+
+
+# TODO
+"""
+│   571 │   │   def show_result(msg: InjectorState):                                               │
+│   572 │   │   │   self.message_broker.unsubscribe(show_result)                                   │
+│ ❱ 573 │   │   │   assert msg.state == STOPPED
+if this fails, then fix the inconsistency instead of crashing
+"""
+
+# TODO test
+
+
+# TODO high-res wheel as input
+# TODO high-res wheel as output
+
+
+
+class RelToRelHandler(MappingHandler):
+    """Handler which transforms EV_REL to EV_REL events"""
+
+    _input_movement: Tuple[int, int]  # (type, code) of the relative movement we map
+    _output_axis: Tuple[int, int]  # the (type, code) of the output axis
+    _transform: Transformation
+    _wheel_input: bool
+
+    # infinite loop which centers the output when input stops
+    _recenter_loop: Optional[asyncio.Task]
+    _moving: asyncio.Event  # event to notify the _recenter_loop
+
+    _remainder: int
+
+    def __init__(
+        self,
+        combination: EventCombination,
+        mapping: Mapping,
+        **_,
+    ) -> None:
+        super().__init__(combination, mapping)
+
+        self._remainder = 0
+
+        self._transform = Transformation(
+            max_=256,
+            min_=-256,
```

Comment:
```suggestion
            max_=mapping.rel_input_cutoff,
            min_=-mapping.rel_input_cutoff,
```

Reply:
or is there a specific reason why you used +-256?

Reply:
Thanks.

Here is an example for when I set my mouse to the fastest speed and move it really fast:

```
Event: time 1666043095.466654, type 2 (EV_REL), code 0 (REL_X), value -243
Event: time 1666043095.466654, type 2 (EV_REL), code 1 (REL_Y), value -16
Event: time 1666043095.466654, -------------- SYN_REPORT ------------
Event: time 1666043095.474654, type 2 (EV_REL), code 0 (REL_X), value -394
Event: time 1666043095.474654, type 2 (EV_REL), code 1 (REL_Y), value -52
Event: time 1666043095.474654, -------------- SYN_REPORT ------------
Event: time 1666043095.482655, type 2 (EV_REL), code 0 (REL_X), value -378
Event: time 1666043095.482655, type 2 (EV_REL), code 1 (REL_Y), value -40
Event: time 1666043095.482655, -------------- SYN_REPORT ------------
Event: time 1666043095.490655, type 2 (EV_REL), code 0 (REL_X), value -370
Event: time 1666043095.490655, type 2 (EV_REL), code 1 (REL_Y), value -31
Event: time 1666043095.490655, -------------- SYN_REPORT ------------
Event: time 1666043095.498654, type 2 (EV_REL), code 0 (REL_X), value -356
Event: time 1666043095.498654, type 2 (EV_REL), code 1 (REL_Y), value -27
Event: time 1666043095.498654, -------------- SYN_REPORT ------------
Event: time 1666043095.506654, type 2 (EV_REL), code 0 (REL_X), value -326
Event: time 1666043095.506654, type 2 (EV_REL), code 1 (REL_Y), value -20
Event: time 1666043095.506654, -------------- SYN_REPORT ------------
Event: time 1666043095.514654, type 2 (EV_REL), code 0 (REL_X), value -354
Event: time 1666043095.514654, type 2 (EV_REL), code 1 (REL_Y), value -25
```

It's not hard at all to produce values higher than 100.

I was about to suggest to increase `rel_input_cutoff`, but it seems it is not really a cutoff. Decreasing `rel_input_cutoff` would make the values from Transformation exceed the value of 1, if the input value is high enough, right? So it is more like an inverse-speed factor

Reply:
No, it is not really a cutoff `Transformation` will happily create values beyond +-1.

For the `RelToRelHandler` It just scales the transformation to a reasonable input range, such that the deadzone and the non-linearity are not to small or to wide.

But more importantly for the `RelToAbsHandler` it is a cutoff and it ensures we can reach +-100% on the ABS output of the mapping. I chose 100 as the default because for me and my mouse that seems to be reasonable speed.

I think we need to chose the default cutoff such that the `RelToAbsHandler` has a good sensitivity for +-100% output. And just make sure the deadzone setting for `RelToRelHandler` feels ok.